platform/linux-generic/odp_ml.c

```diff
@@ -0,0 +1,2673 @@
+
+	if (odp_global_ro.disable.ml) {
+		_ODP_PRINT("ML is disabled\n");
+		return 0;
+	}
+
+	capa->max_model_size = ML_MAX_MODEL_SIZE;
+	capa->max_models = ML_MAX_MODELS_CREATED;
+	capa->max_models_loaded = ML_MAX_MODELS_LOADED;
+	capa->max_compl_id = ML_MAX_COMPL_ID;
+	capa->max_inputs = CONFIG_ML_MAX_INPUTS;
+	capa->max_outputs = CONFIG_ML_MAX_OUTPUTS;
+	capa->max_segs_per_input = ML_MAX_IO_SEGS;
+	capa->max_segs_per_output = ML_MAX_IO_SEGS;
+	capa->min_input_align = 1;
+	capa->min_output_align = 1;
+
+	capa->load.compl_mode_mask = ODP_ML_COMPL_MODE_SYNC |
+				     ODP_ML_COMPL_MODE_POLL |
+				     ODP_ML_COMPL_MODE_EVENT;
+	capa->load.compl_queue_plain = 1;
+	capa->load.compl_queue_sched = 1;
+
+	capa->run.compl_mode_mask =  ODP_ML_COMPL_MODE_SYNC |
+				     ODP_ML_COMPL_MODE_POLL |
+				     ODP_ML_COMPL_MODE_EVENT;
+	capa->run.compl_queue_plain = 1;
+	capa->run.compl_queue_sched = 1;
+
+	if (odp_pool_capability(&pool_capa)) {
+		_ODP_ERR("Pool capability failed\n");
+		return -1;
+	}
+
+	capa->pool.max_pools = pool_capa.buf.max_pools;
+	capa->pool.max_num = pool_capa.buf.max_num;
+	capa->pool.max_uarea_size = pool_capa.buf.max_uarea_size;
+	capa->pool.uarea_persistence = pool_capa.buf.uarea_persistence;
+	capa->pool.max_cache_size = pool_capa.buf.max_cache_size;
+	capa->pool.min_cache_size = pool_capa.buf.min_cache_size;
+
+	return 0;
+}
+
+void odp_ml_config_init(odp_ml_config_t *config)
+{
+	memset(config, 0, sizeof(odp_ml_config_t));
+	config->max_models_created = 1;
+	config->max_models_loaded = 1;
+}
+
+int odp_ml_config(const odp_ml_config_t *config)
+{
+	odp_ml_capability_t *capa = &_odp_ml_glb->capa;
+
+	if (!config) {
+		_ODP_ERR("Error: config must not be NULL\n");
+		return -1;
+	}
+
+	if (config->max_model_size == 0 || config->max_models_created == 0 ||
+	    config->max_models_loaded == 0) {
+		_ODP_ERR("Error: max_model_size, max_models_created and max_models_loaded"
+			 " must be bigger than 0\n");
+		return -1;
+	}
+
+	if (config->max_models_loaded > config->max_models_created) {
+		_ODP_ERR("Error: max_models_loaded %d exceeds max_models_created %d\n",
+			 config->max_models_loaded, config->max_models_created);
+		return -1;
+	}
+
+	if (config->max_models_created > ML_MAX_MODELS_CREATED) {
+		_ODP_ERR("Error: max_models_created %d exceeds maximum number"
+			 " of models that can be created in this driver %d\n",
+			 config->max_models_created, ML_MAX_MODELS_CREATED);
+		return -1;
+	}
+
+	if (config->max_models_loaded > ML_MAX_MODELS_LOADED) {
+		_ODP_ERR("Error: max_models_loaded %d exceeds maximum number"
+			 " of models that can be loaded in this driver %d\n",
+			 config->max_models_loaded, ML_MAX_MODELS_LOADED);
+		return -1;
+	}
+
+	if (config->max_model_size > ML_MAX_MODEL_SIZE) {
+		_ODP_ERR("max_model_size %" PRIu64 " exceeds supported maximum model size %d\n",
+			 config->max_model_size, ML_MAX_MODEL_SIZE);
+		return -1;
+	}
+
+	if (!(capa->load.compl_mode_mask & config->load_mode_mask)) {
```

Comment:
This check should be: (capa->mask & config->mask) == config->mask

I.e. all requested modes are also in capa, now it passes when at least one requested mode is in capa (but others not). Alternatively, could remove the check since all modes are supported.

Reply:
Removed.